File: Permutation.py
from Combination import Combinatorial

# define a class
class Transposition(Combinatorial):

    def __init__(self):
        super().__init__()

    # till the ground for shuffle to grind on
    def possibleWordPermutations(self, candidates, size):
        self.perm_store = [] # list to hold answers
        self.possibleWordCombinations(candidates, size)
        
        # illegal 'r' value
        if len(self.comb_store) == 0 or self.r == 1:
            self.perm_store = self.comb_store
        else:
            last_two = [[], []]
            for i in range(len(self.comb_store)):
                self.index = self.r - 1
                # copy up last two elements of 'comb_store(i)'
                last_two[0] = [self.comb_store[i][self.index], self.comb_store[i][self.index - 1]]
                last_two[1] = last_two[0][::-1] # reverse last_two[0]
                self.index -= 2

                self.local_store = []
                self.local_store.append(tuple(last_two[0]))
                self.local_store.append(tuple(last_two[1]))
                if self.r > 2:
                    self.shuffleWord(self.local_store, i)
                
                self.perm_store.extend(self.local_store)

        return self.perm_store

    def shuffleWord(self, arg_store, i):
        self.local_store = []
        for option in (arg_store):
            members = []
            members.extend(list(option))
            # add 'index' 'comb_store[i]' element to this list of members
            members.append(self.comb_store[i][self.index])

            shift_index = len(members)
            # shuffle this pack of words
            while shift_index > 0:
                # skip if already in store
                if (tuple(members) in self.local_store) == False:
                    self.local_store.append(tuple(members))
                
                shift_index -= 1
                if shift_index > 0 and members[shift_index] != members[shift_index - 1]:
                    # interchange these two neighbours
                    members[shift_index - 1], members[shift_index] = members[shift_index], members[shift_index - 1]
                    
        # Are there any elements left? repeat if yes
        if self.index > 0:
            self.index -= 1
            self.shuffleWord(self.local_store, i)

File: Permutation.js
var origin = [];
var local_store = [];
var perm_store = [];

// till the ground for shuffle to grind on
function possibleWordPermutations(candidates, size) {
    var combination = possibleWordCombinations(candidates, size);
    // illegal 'r' value
    if (combination[0] === undefined || r == 1) {
        perm_store = combination;
    } else {
        for (var i in combination) {
            origin = combination[i].split(separator);
            local_store = [];
            // store up last two words of this element
            local_store.unshift(origin.slice(-2, -1) + separator + origin.slice(-1));
            // store up last two words interchanged;
            // leaving 'origin' without its last 2 elements
            local_store.unshift(origin.pop() + separator + origin.pop());

            if (r > 2) {
                shuffleWord(local_store.slice(0));
            }
            perm_store = perm_store.concat(local_store);
        }
    }
    return perm_store;
}

function shuffleWord(temp_store) {
    local_store = [];
    for (var i in temp_store) {
        var members = temp_store[i].split(separator);
        // add last 'origin' element to this list of members
        members.push(origin.slice(-1));

        var shift_index = members.length;
        var temp_char;
        // shuffle this pack of words
        while (shift_index > 0) {
            var candidate = members.join(separator);
            var find = local_store.indexOf(candidate);
            // skip if already in store
            if (find == -1) {
                local_store.push(candidate);
            }
            shift_index--;
            // interchange these two neighbours
            if (shift_index > 0 && members[shift_index] != members[shift_index - 1]) {
                temp_char = members[shift_index];
                members[shift_index] = members[shift_index - 1];
                members[shift_index - 1] = temp_char;
            }
        }
    }
    // Leave 'origin' without its current last element
    origin.pop();
    // Are there any elements left in origin?
    // if yes, then repeat
    if (origin.length > 0) {
        shuffleWord(local_store);
    }
}
